File: rust/tcl-registry/src/commands/irules/peer.rs
// tcl-lsp — a language server and toolchain for Tcl
// Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! `peer` iRules command.
use crate::prelude::*;
use tcl_dialect::model::SpecSurface;
pub const fn spec() -> CommandSpec {
    CommandSpec {
        name: "peer",
        traits: Traits::IS_SIDE_SWITCH,
        side_switch_target: Some(SideSwitchTarget::Peer),
        surface: Some(SpecSurface::IRULES),
        // `peer NESTING_SCRIPT` — unlike clientside/serverside, peer has
        // no bare query form, so the script body is required: exactly
        // one argument at index 0 (#501).
        arity: Arity::new(1, 1),
        // The nesting script (index 0) is a body evaluated under the
        // peer-side context; it runs synchronously in the caller's
        // frame, so the default `BodyKind::Plain` applies.
        arg_roles: &[(0, ArgRole::Body)],
        hover: Some(HoverSnippet {
            summary: "Causes the specified iRule commands to be evaluated under the peer-side context.",
            synopsis: &["peer ANY_CHARS"],
            snippet: "Causes the specified iRule commands to be evaluated under the peer-side context.",
            source: "https://clouddocs.f5.com/api/irules/peer.html",
            examples: "when SERVER_CONNECTED {\n  peer { TCP::collect }\n}",
            return_value: "",
        }),
        forms: &[FormSpec {
            synopsis: "peer ANY_CHARS",
            ..FormSpec::DEFAULT
        }],
        side_effects: &[SideEffect {
            target: SideEffectTarget::TcpState,
            reads: true,
            connection_side: ConnectionSide::Both,
            ..SideEffect::DEFAULT
        }],
        ..CommandSpec::DEFAULT
    }
}

File: dialects/f5/irules/peer.py
# tcl-lsp — a language server and toolchain for Tcl
# Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: AGPL-3.0-or-later

# Enriched from F5 iRules reference documentation.
"""peer -- Causes the specified iRule commands to be evaluated under the peer-side context."""

from __future__ import annotations

from compiler.registry._base import CommandDef
from compiler.registry.models import CommandSpec, FormKind, FormSpec, HoverSnippet, ValidationSpec
from compiler.registry.namespace_models import EventRequires
from compiler.registry.signatures import ArgRole, Arity
from compiler.side_effects import ConnectionSide, SideEffect, SideEffectTarget

from ._base import _IRULES_ONLY, register

_SOURCE = "https://clouddocs.f5.com/api/irules/peer.html"


def _peer_arg_roles(args: list[str]) -> dict[int, frozenset[ArgRole]]:
    """The nesting script is a body evaluated in the peer-side context.

    ``peer`` is the third side-switch (alongside ``clientside`` /
    ``serverside``); the script is the sole argument (index 0) and runs
    synchronously in the caller's scope (``BodyKind.INLINE``).
    """
    if args:
        return {0: frozenset({ArgRole.BODY})}
    return {}


@register
class PeerCommand(CommandDef):
    name = "peer"

    @classmethod
    def spec(cls) -> CommandSpec:
        return CommandSpec(
            name="peer",
            dialects=_IRULES_ONLY,
            hover=HoverSnippet(
                summary="Causes the specified iRule commands to be evaluated under the peer-side context.",
                synopsis=("peer NESTING_SCRIPT",),
                snippet="Causes the specified iRule commands to be evaluated under the peer-side context.",
                source=_SOURCE,
                examples=("when SERVER_CONNECTED {\n  peer { TCP::collect }\n}"),
            ),
            forms=(
                FormSpec(
                    kind=FormKind.DEFAULT,
                    synopsis="peer NESTING_SCRIPT",
                ),
            ),
            validation=ValidationSpec(
                # ``peer NESTING_SCRIPT`` — unlike clientside/serverside, peer
                # has no bare query form, so the script body is required:
                # exactly one argument at index 0.
                arity=Arity(1, 1),
            ),
            arg_role_resolver=_peer_arg_roles,
            is_side_switch=True,
            event_requires=EventRequires(),
            side_effect_hints=(
                SideEffect(
                    target=SideEffectTarget.TCP_STATE,
                    reads=True,
                    connection_side=ConnectionSide.BOTH,
                ),
            ),
        )
